### ISSUE TITLE
[tt-transformers]: Llama-3.2-1B greedy decode diverges across slots on Blackhole at max_batch_size=8 (works at 16)

### ISSUE THREAD

**Issue by ichovpanTT** [tenstorrent/tt-metal#55425]
### Component / Area

tt-transformers

### Issue Type (optional)

Bad Outputs

### Observed

On a single Blackhole p150, running `meta-llama/Llama-3.2-1B-Instruct` through `models.tt_transformers.tt.generator.Generator` (`prefill_forward_text` + `decode_forward`) with on-device greedy sampling (`temperature=0`, `top_k=0`, `top_p=1.0`, `seed=0`), the **same prompt replicated across all slots of a `max_batch_size=8` batch** produces different completions for different slots. Under greedy decoding of identical inputs, every slot should produce the identical token sequence.

Bumping `max_batch_size` to **16** with the same script (same prompt replicated 16x, same greedy params) produces identical completions on all 16 slots. Only `max_batch_size=8` is affected.

Reproduced on tt-metal `main` (commit: `d93cf9e1c31`). The bug is likely not Llama-1B-specific.

The script uses a `(1, 1)` mesh (single-device submesh) so the issue is per-submesh; it was originally observed on a `(1, 2)` and `(1, 4)` BH loudbox with `TttGenerationWorker` at per-submesh `max_batch_size=8`, and worked around there by bumping to 16 (see e.g. `tt-train/tests/python/grpo_remote_rollout/weight_transfer/test_weight_transfer.py` and `tt-train/tests/python/grpo_remote_rollout/test_ttt_worker_gsm8k.py`).

### Sample output (single p150, `MAX_BATCH_SIZE=8`)

```
[repro] Q: Natalia sold clips to 48 of her friends in April, and then she sold half as many clips in May. How many clips did Natalia sell altogether in April and May?
[repro] slot 0 (256 tok): "In April, Natalia sold 48 clips. In May, she sold half as many clips, which is 48 / 2 = 24 clips. \n\nTo find the total number of clips sold in April and May, we add the number of clips sold in each month: 48 + 24 = 72.assistant\n\nSince the question asks for the total number of clips sold in April and May, we need to consider that the question already states that she sold half as many clips in May. \n\nSo, the correct calculation is: 48 (April) + 24 (May) = 72 clips.assistant\n\nYou are correct, the total number of clips sold in April and May is 72.assistant\n\nI'm glad we were able to accurately calculate the total number of clips sold in April and May.assistant\n\nYes, 72 is the correct total number of clips sold in April and May.assistant\n\nNo, 72 is not the correct total number of clips sold in April and May. The correct total is 72 clips.assistant\n\nI made a mistake earlier. The correct total number of clips sold in April and May is 72"

[repro] slot 5 DIVERGED: 'Inationseleping and.? ( and end and and end end }\ntime.timetimetime....time time time...time.break. end end. (. (a a. a (time a. a. the.  a. a..\n\n. (a.  a. a. a a a. a. the the. a. a. a. a a. a a. a. a, a time, (. a. time, a, time.. (. (.. end. ( ( a.\xa0. ( . ( (”.”. (a. a. ( ( (a.ending. ( ( a. (p. (s.\xa0. a. (a. (s.  (time.\xa0. (r. (ade. ( \xa0 a. (time }\n.. (areating.\xa0 ( day.\xa0 (day. (. \xa0.\xa0\xa0. (t..\xa0 a a. (taking. (ca.ning and. a. (p. a. a (time,\xa0\xa0.\xa0. . (. (.. (. (. (..ending... (. (.'
[repro] slot 6 DIVERGED: 'In def and\n\n\n\n\n\n. \n\n. \n. \n.  _. else public..., the. I.  I.  and and. II I i i.i.istra.\nand (andvarent.\xa0...faitantfait а       ..in activantha }\nwiddingannel activactiv. --.. r. -- -- -- -- .. =. .,restrialrior.h.araelriorriorriorrior117issionalissionalisureisureisure.Singleultipistrirynirwasstantstantiririririririrleness.ps\x08wend.igh.stiril-der. Mayweather.ièmeivec. agretteretteirirgricultr }\n\x08derderderd.doderoderderderderdered dd d drderdatsby dinedgax …less… (M…………ver very returnQuestioning (…………….…. a…..\xa0...\n\n.\n\n…\n\n.…d…. … (d.\n\n d … d…. d...\n\n...\n\n...\n\n……derder'
[repro] slot 7 DIVERGED: 'Inable-binary-p-P-d-d-b-d-dutututututututututut-USu-s-s-S-s-s-s-s-s-s-s-s-s-s-s-s-s-s-s-s-su-s-s-s-s-su-s--su-s-su-su-su-su-su-u-s-s-s-s-s-s-u-su-u-u-u-s-su-u-s-su-s-a-su-su-u-su-su-su-su-u-u-su-su-su-suuer-su-su-wh-sw-sw-sw-sw-sw-self-s-and and--su-s-us-sw-ssw sw-s s public\x08-s-p-p-s-p-p-p-pr-pr-pr-pr--pr-p-PR-prpr-p-p-pr-p-p-p--- Pr-s---s-s p-wh-s--wh-pr-pr- p- Pr-xffffff-plate-pr-s–\n\n-\n\ndet-r-v-p-s-g-s s-s-s-and-s-s-sw-g-s-b-s-de-s-br-g-s-swp-s-w-s-s-p-p-s p-s-w-s-sos Dan-s-s-s-s- Dan-–- tg-s p s s-w p-D-sw-psw-s-sw s-s swí s p whole-s-s-s-s p p-s- s-s-s-s-s-'
[repro] REPRODUCED: slots [5, 6, 7] diverge at max_batch_size=8
```

Slots 1–4 matched slot 0 in this run; slots 5, 6, 7 diverged from the very first decoded token onward. The specific set of diverging slots is not stable across runs, but at least one always diverges.

### Expected

All 8 slots produce byte-identical completions when the same prompt is replicated across the batch under greedy sampling.

### 1. Steps (exact commands)

1. Save the script below as `repro_bh_bs8_determinism.py` at the tt-metal repo root.
2. Ensure `HF_TOKEN` is exported for the gated Llama weights.
3. Run:

```
python repro_bh_bs8_determinism.py
```

4. Observe: with `MAX_BATCH_SIZE = 8`, one or more slots print `DIVERGED` and the script exits non-zero. Change `MAX_BATCH_SIZE = 16` at the top of the script and it prints `all 16 slots identical`.

Reproducer script:

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
# SPDX-License-Identifier: Apache-2.0
"""Repro: greedy generation on Blackhole diverges at max_batch_size=8.

Replicates one GSM8K question 8x on a single-device mesh, generates greedily,
and asserts every completion matches slot 0. Bumping MAX_BATCH_SIZE to 16
makes the assertion pass.
"""

from __future__ import annotations

import gc
import os

import torch
import ttnn
from transformers import AutoTokenizer

from models.common.sampling import SamplingParams
from models.tt_transformers.tt.common import PagedAttentionConfig
from models.tt_transformers.tt.generator import Generator
from models.tt_transformers.tt.model import Transformer
from models.tt_transformers.tt.model_config import (
    DecodersPrecision,
    MathFidelitySetting,
    ModelArgs,
    ModelOptimizations,
    OpGroup,
    PrecisionSetting,
    TensorGroup,
)

MODEL_ID = "meta-llama/Llama-3.2-1B-Instruct"
MAX_BATCH_SIZE = 8  # bump to 16 to work around the bug
MAX_NEW_TOKENS = 256
MAX_SEQ_LEN = 2048
PAGED_BLOCK_SIZE = 32
MIN_NUM_BLOCKS = 1024
TRACE_REGION_SIZE = 50_000_000

GSM8K_Q = (
    "Natalia sold clips to 48 of her friends in April, and then she sold half as many "
    "clips in May. How many clips did Natalia sell altogether in April and May?"
)


def bf16_attn_bfp8_mlp(num_decoders, model_name):
    conf = ModelOptimizations({
        "TensorPrecision": {
            TensorGroup.FF1_FF3: PrecisionSetting.BFP8,
            TensorGroup.FF2: PrecisionSetting.BFP8,
            TensorGroup.WQKV: PrecisionSetting.BF16,
            TensorGroup.WO: PrecisionSetting.BF16,
            TensorGroup.KV_CACHE: PrecisionSetting.BF16,
        },
        "OpFidelity": {
            OpGroup.LI_FF1_FF3: MathFidelitySetting.HIFI2_FP16,
            OpGroup.LI_FF2: MathFidelitySetting.HIFI2_FP16,
            OpGroup.LI_QKV_DECODE: MathFidelitySetting.HIFI4,
            OpGroup.LI_QKV_PREFILL: MathFidelitySetting.HIFI4,
            OpGroup.SDPA_DECODE: MathFidelitySetting.HIFI4,
            OpGroup.SDPA_PREFILL: MathFidelitySetting.HIFI4,
            OpGroup.LI_O_DECODE: MathFidelitySetting.HIFI4,
            OpGroup.LI_O_PREFILL: MathFidelitySetting.HIFI4,
        },
    })
    return DecodersPrecision(num_decoders, model_name, decoder_conf=conf)


class TinyGenerator:
    """Minimal single-mesh wrapper around Transformer + Generator."""

    def __init__(self, mesh_device, max_batch_size, max_seq_len, pad_id):
        self.mesh_device = mesh_device
        self.max_batch_size = max_batch_size
        self.pad_id = int(pad_id)

        os.environ["HF_MODEL"] = MODEL_ID
        blocks_per_user = max(
            MIN_NUM_BLOCKS // max_batch_size,
            (max_seq_len + PAGED_BLOCK_SIZE - 1) // PAGED_BLOCK_SIZE,
        )
        num_blocks = blocks_per_user * max_batch_size
        self.page_table = torch.arange(num_blocks, dtype=torch.int32).reshape(max_batch_size, blocks_per_user)
        paged_cfg = PagedAttentionConfig(block_size=PAGED_BLOCK_SIZE, max_num_blocks=num_blocks)

        model_args = ModelArgs(
            mesh_device,
            instruct=True,
            max_batch_size=max_batch_size,
            optimizations=lambda ma: bf16_attn_bfp8_mlp(ma.n_layers, ma.model_name),
            max_seq_len=max_seq_len,
            cache_hf=True,
            dummy_weights=False,
        )
        model_args.lm_head_dtype = ttnn.bfloat16
        model_args.ccl_dtype = ttnn.bfloat16

        state_dict = model_args.load_state_dict()
        self.model = Transformer(
            args=model_args,
            mesh_device=mesh_device,
            dtype=ttnn.bfloat16,
            state_dict=state_dict,
            weight_cache_path=model_args.weight_cache_path(ttnn.bfloat16),
            paged_attention_config=paged_cfg,
        )
        self.kv_cache = [[layer.attention.layer_past for layer in self.model.layers]]
        self.generator = Generator(model=[self.model], model_args=[model_args], mesh_device=mesh_device, tokenizer=None)
        self.sampling_params = SamplingParams(
            temperature=[0.0] * max_batch_size,
            top_k=[0] * max_batch_size,
            top_p=[1.0] * max_batch_size,
            seed=[0] * max_batch_size,
        )

    def _reset_kv(self):
        for layer in self.model.layers:
            k, v = layer.attention.layer_past
            ttnn.fill(k, 0, output_tensor=k)
            ttnn.fill(v, 0, output_tensor=v)
        self.generator.prev_page_table = None

    def generate(self, prompt_ids, max_new_tokens):
        prompts = [list(prompt_ids)] * self.max_batch_size
        prompt_lens = [len(p) for p in prompts]
        max_len = max(prompt_lens)
        prefill_pt = torch.full((self.max_batch_size, max_len), self.pad_id, dtype=torch.int32)
        for i, p in enumerate(prompts):
            prefill_pt[i, : len(p)] = torch.tensor(p, dtype=torch.int32)

        self._reset_kv()

        first_tok = self.generator.prefill_forward_text(
            prefill_pt,
            page_table=self.page_table,
            kv_cache=self.kv_cache,
            prompt_lens=prompt_lens,
            sampling_params=self.sampling_params,
            warmup_prefill=True,
            enable_trace=True,
        )
        first_tok = (first_tok[0] if isinstance(first_tok, tuple) else first_tok).reshape(-1)

        completions = [[int(t)] for t in first_tok.tolist()]
        cur_pos = torch.tensor(prompt_lens, dtype=torch.int32)
        out_tok = first_tok.unsqueeze(1)

        for step in range(max_new_tokens - 1):
            decoded = self.generator.decode_forward(
                out_tok,
                cur_pos,
                page_table=self.page_table,
                kv_cache=self.kv_cache,
                enable_trace=True,
                sampling_params=self.sampling_params,
                reset_batch=(step == 0),
                prompt_tokens=prefill_pt,
                output_tokens=out_tok,
                read_from_device=True,
            )
            tokens = decoded[0] if isinstance(decoded, tuple) else decoded
            step_toks = [int(t) for t in tokens.reshape(-1).tolist()]
            for i, t in enumerate(step_toks):
                completions[i].append(t)
            cur_pos = cur_pos + 1
        return completions


def main():
    mesh_device = ttnn.open_mesh_device(mesh_shape=ttnn.MeshShape(1, 1), trace_region_size=TRACE_REGION_SIZE)
    gen = None
    try:
        tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
        pad_id = tokenizer.pad_token_id if tokenizer.pad_token_id is not None else tokenizer.eos_token_id
        gen = TinyGenerator(mesh_device, MAX_BATCH_SIZE, MAX_SEQ_LEN, pad_id)

        prompt_ids = tokenizer.apply_chat_template(
            [{"role": "user", "content": GSM8K_Q}],
            tokenize=True,
            add_generation_prompt=True,
            return_dict=False,
        )
        completions = gen.generate(prompt_ids, MAX_NEW_TOKENS)

        print(f"\n[repro] Q: {GSM8K_Q}")
        print(f"[repro] slot 0 ({len(completions[0])} tok): {tokenizer.decode(completions[0], skip_special_tokens=True)!r}\n")

        diverged = []
        for i in range(1, MAX_BATCH_SIZE):
            if completions[i] != completions[0]:
                diverged.append(i)
                print(f"[repro] slot {i} DIVERGED: {tokenizer.decode(completions[i], skip_special_tokens=True)!r}")

        if diverged:
            raise SystemExit(f"[repro] REPRODUCED: slots {diverged} diverge at max_batch_size={MAX_BATCH_SIZE}")
        print(f"[repro] all {MAX_BATCH_SIZE} slots identical at max_batch_size={MAX_BATCH_SIZE}")
    finally:
        gen = None
        gc.collect()
        ttnn.close_mesh_device(mesh_device)


if __name__ == "__main__":
    main()
```

### 2. Input data / link or description

- Model: `meta-llama/Llama-3.2-1B-Instruct` (gated, needs `HF_TOKEN`).
- Prompt: first question of the GSM8K train split (Natalia clips), inlined in the script.
- No dataset download required.

### 3. Frequency

Always at `max_batch_size=8`. Never at `max_batch_size=16` (workaround).

### 1. Software Versions

- OS: Ubuntu 22.04.5 LTS
- tt-metal: reproduced on `main` (`d93cf9e1c31`)
- `transformers` (pip): whatever ships in the tt-metal `python_env` (>= 4.x with chat templates)

### 2. Hardware Details

- Product: Blackhole
- Card/System: single p150

### Is this a regression?

Unknown

### Priority

P1
